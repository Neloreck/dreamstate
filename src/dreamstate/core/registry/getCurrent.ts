import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyObject, TDreamstateService } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TDreamstateService<any>>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
