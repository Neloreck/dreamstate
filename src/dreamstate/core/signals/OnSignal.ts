import { SIGNAL_METADATA_REGISTRY } from "@/dreamstate/core/internals";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { TAnyContextManagerConstructor, TSignalType } from "@/dreamstate/types";
import { createMethodDecorator } from "@/dreamstate/utils/polyfills/createMethodDecorator";

/**
 * Class method decorator.
 * Marks decorated method as handler of provided type(s) signals.
 * All signals in current scope with specified type will be handled by callback.
 *
 * @param {(TSignalType|Array.<TSignalType>>)} signalType - signal or array of signals that should be handled.
 */
export function OnSignal(signalType: Array<TSignalType> | TSignalType): MethodDecorator {
  if (!signalType) {
    throw new TypeError("Signal type should be provided for OnQuery decorator.");
  }

  /**
   * Support old and new decorators with polyfill.
   */
  return createMethodDecorator<TAnyContextManagerConstructor>(function(
    method: string | symbol,
    ManagerClass: TAnyContextManagerConstructor
  ): void {
    if (!(ManagerClass.prototype instanceof ContextManager)) {
      throw new TypeError("Only ContextManager extending classes methods can be decorated as handlers.");
    }

    if (SIGNAL_METADATA_REGISTRY.has(ManagerClass)) {
      SIGNAL_METADATA_REGISTRY.get(ManagerClass)!.push([ method, signalType ]);
    } else {
      SIGNAL_METADATA_REGISTRY.set(ManagerClass, [ [ method, signalType ] ]);
    }
  });
}
