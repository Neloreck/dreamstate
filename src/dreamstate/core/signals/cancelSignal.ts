import { ISignalEvent } from "@/dreamstate/types";

export function cancelSignal(this: ISignalEvent) {
  this.canceled = true;
}
