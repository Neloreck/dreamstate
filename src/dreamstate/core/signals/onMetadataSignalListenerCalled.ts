import { CONTEXT_SIGNAL_METADATA_REGISTRY } from "@/dreamstate/core/internals";
import { ISignalEvent, TConstructorKey, TDreamstateService, TSignalType } from "@/dreamstate/types";

/**
 * Listen signal and call related metadata listeners of this manager.
 */
export function onMetadataSignalListenerCalled<T extends TSignalType = TSignalType, D = undefined>(
  this: InstanceType<TDreamstateService>,
  signal: ISignalEvent<T, D>
): void {
  /**
   * Ignore managers without metadata.
   */
  if (CONTEXT_SIGNAL_METADATA_REGISTRY.has(this.constructor as TDreamstateService)) {
    for (const [ method, subscribed ] of CONTEXT_SIGNAL_METADATA_REGISTRY.get(this.constructor as TConstructorKey)!) {
      if (Array.isArray(subscribed) ? subscribed.includes(signal.type) : signal.type === subscribed) {
        (this as any)[method](signal);
      }
    }
  }
}
