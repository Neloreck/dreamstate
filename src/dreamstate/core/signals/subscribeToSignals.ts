import { SIGNAL_LISTENERS_REGISTRY } from "@/dreamstate/core/internals";
import { TSignalListener, TSignalType } from "@/dreamstate/types";

/**
 * Subscribe to all signals and listen everything.
 * Should be filtered by users like redux does.
 * Not intended to be used as core feature, just for some elegant decisions.
 *
 * Returns function that unsubscribe current handler.
 */
export function subscribeToSignals(
  listener: TSignalListener<TSignalType, any>
): () => void {
  if (typeof listener !== "function") {
    throw new Error(`Signal listener must be function, '${typeof listener}' provided.`);
  }

  SIGNAL_LISTENERS_REGISTRY.add(listener);

  return function() {
    SIGNAL_LISTENERS_REGISTRY.delete(listener);
  };
}
