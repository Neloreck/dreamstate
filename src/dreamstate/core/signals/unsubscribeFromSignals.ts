import { SIGNAL_LISTENERS_REGISTRY } from "@/dreamstate/core/internals";
import { TSignalListener } from "@/dreamstate/types";

/**
 * Unsubscribe from all signals and listen everything.
 * Not intended to be used as core feature, just for some elegant decisions.
 */
export function unsubscribeFromSignals(listener: TSignalListener<any, any>): void {
  SIGNAL_LISTENERS_REGISTRY.delete(listener);
}
