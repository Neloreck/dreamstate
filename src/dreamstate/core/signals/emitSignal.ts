import { log } from "@/macroses/log.macro";

import { IRegistry } from "@/dreamstate/core/scoping/registry/createRegistry";
import {
  IBaseSignal,
  ISignalEvent,
  TAnyContextManagerConstructor,
  TSignalListener
} from "@/dreamstate/types";

/**
 * Callback for signal canceling.
 * Used in signal events for propagation stopping.
 */
function cancelSignal<D = undefined>(this: ISignalEvent<D>): ISignalEvent<D> {
  this.canceled = true;

  return this;
}

/**
 * Emit signal and notify all subscribers in an async way.
 * Composes signal event from base data and notifies all listeners in current scope.
 * If event is canceled, stop its propagation to next handlers.
 */
export function emitSignal<D = undefined>(
  base: IBaseSignal<D>,
  emitter: TAnyContextManagerConstructor | null = null,
  REGISTRY: IRegistry
): ISignalEvent<D> {
  if (!base || base.type === undefined) {
    throw new TypeError("Signal must be an object with declared type.");
  }

  const signalEvent: ISignalEvent<D> = {
    type: base.type,
    data: base.data as D,
    emitter,
    timestamp: Date.now(),
    cancel: cancelSignal
  };

  REGISTRY.SIGNAL_LISTENERS_REGISTRY.forEach(function(it: TSignalListener<D>) {
    try {
      if (!signalEvent.canceled) {
        it(signalEvent);
      }
    } catch (error) {
      log.error(`Failed to proceed emitted signal (${String(base.type)}):`, error);
    }
  });

  return signalEvent;
}
