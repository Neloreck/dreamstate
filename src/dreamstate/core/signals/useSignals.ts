import { useEffect } from "react";

import { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
import { TCallable, TSignalListener, TSignalType } from "@/dreamstate/types";

/**
 * Hook for signals listening and custom UI handling.
 * Provided dependencies will trigger function re-subscription.
 */
export function useSignals<
  T extends TSignalType = TSignalType,
  D = undefined
>(
  subscriber: TSignalListener<T, D>,
  dependencies: Array<any>
): void {
  useEffect(function(): TCallable {
    return subscribeToSignals(subscriber);
  }, dependencies);
}
