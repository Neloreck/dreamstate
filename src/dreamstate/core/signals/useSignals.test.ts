import { mount } from "enzyme";
import { createElement } from "react";
import { act } from "react-dom/test-utils";

import { getCurrent } from "@/dreamstate/core/registry/getCurrent";
import { registerService } from "@/dreamstate/test-utils/registry/registerService";
import { unRegisterService } from "@/dreamstate/test-utils/registry/unRegisterService";
import { nextAsyncQueue } from "@/dreamstate/test-utils/utils/nextAsyncQueue";
import { EmittingContextManager, ESignal, TStringSignalEvent, SubscribedFunctionalComponent } from "@/fixtures/signals";

describe("useSignals method", () => {
  beforeEach(() => {
    registerService(EmittingContextManager);
  });

  afterEach(() => {
    unRegisterService(EmittingContextManager);
  });

  it("Functional components should properly subscribe to signals and update ui", async () => {
    const emittingContextManager: EmittingContextManager = getCurrent(EmittingContextManager)!;

    const mockFn = jest.fn((signalEvent: TStringSignalEvent) => {
      expect(signalEvent.type).toBe(ESignal.STRING_SIGNAL);
      expect(signalEvent.emitter).toBe(EmittingContextManager);
      expect(signalEvent.data).toBe("newValue");
    });
    const tree = mount(
      createElement(SubscribedFunctionalComponent, { onInternalSignal: (event: TStringSignalEvent) => mockFn(event) })
    );

    expect(tree).toMatchSnapshot();

    await act(async () => {
      emittingContextManager.sendStringSignal("newValue");

      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(mockFn).toHaveBeenCalled();

    mockFn.mockClear();

    tree.unmount();

    emittingContextManager.sendStringSignal("newValue");
    await nextAsyncQueue();

    expect(mockFn).not.toHaveBeenCalled();
  });
});
