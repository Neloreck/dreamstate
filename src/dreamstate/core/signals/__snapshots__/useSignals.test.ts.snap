// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSignals method. Functional components should properly subscribe to signals and update ui. 1`] = `
<SubscribedFunctionalComponent
  onInternalSignal={[Function]}
>
  <div>
    initial
  </div>
</SubscribedFunctionalComponent>
`;

exports[`useSignals method. Functional components should properly subscribe to signals and update ui. 2`] = `
<SubscribedFunctionalComponent
  onInternalSignal={[Function]}
>
  <div>
    newValue
  </div>
</SubscribedFunctionalComponent>
`;
